controllers/kvmcluster_controller.go

```diff
@@ -0,0 +1,159 @@
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the KVMCluster object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *KVMClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reconcileError error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	defer metrics.CaptureMetrics()
+
+	// Fetch the KVMCluster instance
+	kvmCluster := &v1alpha4.KVMCluster{}
+	err := r.Get(ctx, req.NamespacedName, kvmCluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	defer func() {
+		if reconcileError != nil {
+			r.Recorder.Event(kvmCluster, "Warning", "failed-to-reconcile", reconcileError.Error())
+		}
+	}()
+
+	// Fetch the Cluster.
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, kvmCluster.ObjectMeta)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if cluster == nil {
+		log.Info("Cluster Controller has not yet set OwnerRef")
+		return reconcile.Result{}, nil
+	}
+
+	if annotations.IsPaused(cluster, kvmCluster) {
+		log.Info("KVMCluster or linked Cluster is marked as paused. Won't reconcile")
+		return reconcile.Result{}, nil
+	}
+
+	log = log.WithValues("cluster", cluster.Name)
+
+	// Create the scope.
+	clusterScope, err := scope.NewClusterScope(scope.ClusterScopeParams{
+		Cluster:        cluster,
+		KVMCluster:     kvmCluster,
+		ControllerName: KVMClusterController,
+	})
+	if err != nil {
+		return reconcile.Result{}, errors.Errorf("failed to create scope: %+v", err)
+	}
+
+	// Always close the scope when exiting this function so we can persist any KVMCluster changes.
+	defer func() {
+		if err := clusterScope.Close(); err != nil && reconcileError == nil {
+			reconcileError = err
+		}
+	}()
+
+	// Handle deleted clusters
+	if !kvmCluster.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, clusterScope)
+	}
+
+	// Handle non-deleted clusters
+	return r.reconcileNormal(ctx, clusterScope)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KVMClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha4.KVMCluster{}).
+		Complete(r)
+}
+
+func (r *KVMClusterReconciler) reconcileNormal(ctx context.Context, scope *scope.ClusterScope) (_ reconcile.Result, reconcileError error) {
+
+	// If the KVMCluster doesn't have our finalizer, add it.
+	controllerutil.AddFinalizer(scope.KVMCluster, v1alpha4.ClusterFinalizer)
+
+	// Register the finalizer immediately to avoid orphaning resources on delete
+	if err := scope.PatchObject(); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	nsService := namespace.NewService()
+	nsService.ReconcileNamespace()
```

Comment:
So my understanding is these services should be an analogue to the resources that we currently have in [kvm-operator](https://github.com/giantswarm/kvm-operator/tree/master/service/controller/resource).